caddytls/ctmonitor.go

```diff
@@ -0,0 +1,308 @@
+	}
+	return
+}
+
+// lookUpNames queries the certSpotter service for each SAN that Caddy is hosting
+// It then adds them to a set and returns a map of each certificate  mapped to a string
+// that contains identifying information for the cert.
+func lookUpNames(caddyCertsSANs []string, config QueryConfig) (map[string]string, int, error) {
+	// retrievedCerts is the bytes of a certificate mapped to the
+	// ID, issuer name, and before/after values.
+	retrievedCerts := make(map[string]string)
+
+	// biggestID is the most recent certificate id returned from CertSpotter.
+	// timeToWait is the amount of time you need to wait before querying again.
+	// initialIndex saves the starting point for domainName queries.
+	var biggestID, timeToWait, initialIndex int
+
+	// retryAfter is the timeToWait value from the response headers.
+	var retryAfter string
+
+	var err error
+	initialIndex = config.Index
+
+	// If the caddyCertsSANs is empty, it shouldn't run this at all.
+	for _, domainName := range caddyCertsSANs {
+		config.Index = initialIndex // resets the index for the next query so that no certs are skipped.
+		retryAfter, err = queryDomainName(domainName, config, &biggestID, retrievedCerts)
+		if err != nil {
+			timeToWait = 3600
+			return retrievedCerts, timeToWait, fmt.Errorf("queryDomainName encountered error: %v", err)
+		}
+	}
+	err = putLatestID(biggestID, filePath)
+	if err != nil {
+		timeToWait = 3600
+		return retrievedCerts, timeToWait, fmt.Errorf("[ERROR] writing latest ID failed, err: %v", err)
+	}
+	timeToWait, err = strconv.Atoi(retryAfter)
+	if err != nil {
+		log.Printf("Error %v retrieving time to wait, waiting 1 hour\n", err)
+		timeToWait = 3600
+	}
+	// returning retrievedCerts because it is created here and timeToWait
+	// so that ctmonitor waits enough time before trying again.
+	return retrievedCerts, timeToWait, err
+}
+
+// monitorCerts continuously monitors the certificates that Caddy serves,
+// it queries again after the specified time.
+func monitorCerts() {
+	//var queryConfig QueryConfig
+	config, err := loadConfig()
+	if err != nil {
+		log.Printf("[ERROR] loading config: %v", err)
+		return
+	}
+	for {
+		namesToLookUp, caddyCerts := getCaddyCerts()
+		if len(namesToLookUp) == 0 {
+			log.Printf("[ERROR]Could not retrieve DNS names from Caddy Certificate, " +
+				"make sure that you are serving on port 80 & 443. " +
+				"Terminating monitorCerts.\n")
+		}
+		startingIndex, err := getLatestIndex(filePath)
+		if err != nil {
+			log.Printf("Error %v while getting starting index, starting at 0", err)
+			break
+		}
+		config.Query = certSpotterAPIBase
+		config.Index = startingIndex
+		fetchedCerts, pause, err := lookUpNames(namesToLookUp, config)
+
+		// CompareCerts runs even if there was an error.  If the error occurred after
+		// retrieving all of the certs, we can still compare them before quitting.
+		CompareCerts(caddyCerts, fetchedCerts)
+		if err != nil {
```

Comment:
This `err` value is not assigned before the last check.

Reply:
The comment before explains that it runs even if there was an error, the error would have occurred previously but compareCerts will check any that it has already found.

Reply:
Ok, I took out the new line between the comment and the function before that assigns the err value.